### ISSUE TITLE
Plugin infrastructure

### ISSUE THREAD

**Issue by dune73** [coreruleset/coreruleset#2003]
### Motivation

We decided we are going to provide a plugin structure for CRS. This Issue here is meant to discuss the architecture. I and @azurit volunteered to work on this.

@Azurit has a working proposal in https://github.com/coreruleset/coreruleset/pull/1993.

I'll submit a simpler variant below.

**Comment by dune73** [comment]
# Proposal for plugin structure

## Plugins folder

The plugins folder is residing within CRS on the top level, thus next to the rules folder.

It contains two empty files that will make sure we do not get an Include error (since we can not use IncludeOptional on NGINX). Please note, there is meant to be a single plugins folder for configuration, _early files_ and _late files_ as well.

```
$ ls -1 plugins
empty-early.conf
empty-late.conf
```


## Includes

We are putting

```

# === ModSecurity Core Rule Set Configuration Inclusion

Include     /home/dune73/data/git/crs-mypull-plugin-structure/crs-setup.conf

# === ModSecurity Core Rule Set Inclusion

Include     /home/dune73/data/git/crs-mypull-plugin-structure/plugins/*-early.conf
Include     /home/dune73/data/git/crs-mypull-plugin-structure/rules/*.conf
Include     /home/dune73/data/git/crs-mypull-plugin-structure/plugins/*-late.conf
```

## Example plugin "example"

```
$ cat plugins/example-config-early.conf
SecAction "id:9501000,phase:1,pass,nolog,setvar:TX.example=123"

$ cat plugins/example-early.conf
SecRule ARGS "@streq %{TX.example}" "id:9501100,phase:2,pass,log,msg:'HIT early',setvar:'tx.anomaly_score_pl1=+%{tx.critical_anomaly_score}'"

$ cat plugins/example-late.conf
SecRule ARGS_GET "@streq %{TX.example}" "id:9501110,phase:1,pass,log,msg:'HIT late',setvar:'tx.anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
```

## How it works

The Includes will try and load the plugins rules from the plugins folder. There is a separate Include for rules files that need to be loaded before the CRS include and one to be loaded after the CRS files. The Include work with a wildcard pattern `*-early.conf` and `*-late.conf`. That way the files can reside in the same folder, but they are loaded selectively.

Optional plugin configuration files have a filename of the pattern `<plugin-name>-config-early.conf`. The ending guarantees, the configuration is loaded together with the early rules, but _before_ the early rules thanks to the alphabetic order.

## TODO

* Test on ModSec v3
* Add SecComponentSignature
* Document this more throughly

**Comment by azurit** [comment]
I like it but note, that this isn't going to work:
```
$ cat plugins/example-early.conf
SecRule ARGS "@streq %{TX.example}" "id:9501100,phase:2,pass,log,msg:'HIT early',setvar:'tx.anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
```
Variable `tx.anomaly_score_pl1` is inicialized inside CRS in `REQUEST-901-INITIALIZATION.conf` (my plugin implementation has the same problem - no access to paranoia score variables). We will, probably, need to move the initialization of these variables or move the whole `REQUEST-901-INITIALIZATION.conf` file.

**Comment by dune73** [comment]
It is working, since the rule is in phase 2 and the initialization happens in phase 1. But it is true that you can not do rules for phase:1 in the early config these have to go into the late one. And it's the opposite with rules for phase 2, they need to be in the early file, since 949110 will block the request before it hits the late file.

Therefore: 

phase:1 rules:  <plugin-name>-late.conf
phase:2 rules:  <plugin-name>-early.conf

Correct?

**Comment by azurit** [comment]
Yes but it's really confusing: rules in 'early' are run lately and rules in 'late' are run early.

**Comment by azurit** [comment]
In fact, there are two 'early' and two 'late' stages (in the CRS and modsecurity point of view).

**Comment by dune73** [comment]
I agree with the two stages each and with the confusion nature of this.

With my proposal, I make it very simple for admins (who do not need to understand the exact mechanics), but harder for people developing a plugin. Yet I think we can support those with a dummy plugin they can use as a template for their own ones. And we can document it throughly. With a diagram that explains the order of execution.

**Comment by azurit** [comment]
Ok but maybe we should not use words 'early' and 'late'. Using 'before-crs' and 'after-crs' isn't good too, as when putting rule for phase 1 into 'after-crs', it will be run only after CRS rules for phase 1 BUT before rules for phases 2-4. It's confusing even for me, what about third-party plugin writers? But i can't come up with something better.

Let's talk about ID ranges for plugins. Do we have enough IDs?

**Comment by dune73** [comment]
I agree, better terms would be useful.

IDs: I'd propose the range from 9,500,000 on-wards with 1K per plugin and you register via a PR to a separate plugin register project.

**Comment by azurit** [comment]
Is anything more needs to be discussed here? The proposed implementation is good and should work ok (and it's really simple).

**Comment by dune73** [comment]
I still have not tested this on ModSec3 / NGINX and we probably need to find a replacement for the terms "early" and "late".

But I am going to point the agenda of tonight's meeting to this proposal here.

**Comment by fzipi** [comment]
We use the exactly same scheme, but we normally name "pre" and "post"...

**Comment by dune73** [comment]
Remark during the monthly chat by @airween: We already use _before_ and _after_ for some of our files. May make sense to re-use these terms.

**Comment by dune73** [comment]
Plugin infrastructure PR has been merged (#2038). Closing this.
